libraries/chain/finality/qc.cpp

```diff
@@ -0,0 +1,505 @@
+        ("c", connection_id)("bn", block_num)("sv", strong)("s", s)("pre", pre_state)("state", post_state)("q", is_quorum_met(post_state)));
+   return s;
+}
+
+// called by get_best_qc which acquires a mutex
+qc_sig_t open_qc_sig_t::to_valid_qc_sig() const {
+   qc_sig_t valid_qc_sig;
+
+   if( pending_state == state_t::strong ) {
+      valid_qc_sig.strong_votes = strong_votes.bitset;
+      valid_qc_sig.sig          = strong_votes.sig;
+   } else if (is_quorum_met_no_lock()) {
+      valid_qc_sig.strong_votes = strong_votes.bitset;
+      valid_qc_sig.weak_votes   = weak_votes.bitset;
+      valid_qc_sig.sig          = strong_votes.sig;
+      valid_qc_sig.sig.aggregate(weak_votes.sig);
+   } else
+      assert(0); // this should be called only when we have a valid qc_t.
+
+   return valid_qc_sig;
+}
+
+std::optional<qc_sig_t> open_qc_sig_t::get_best_qc() const {
+   std::lock_guard g(*_mtx);
+   // if this does not have a valid QC, consider received_qc_sig only
+   if( !is_quorum_met_no_lock() ) {
+      if( received_qc_sig ) {
+         return std::optional{*received_qc_sig};
+      }
+      return {};
+   }
+
+   // extract valid QC sig from open qc sig
+   qc_sig_t valid_qc_sig_from_open = to_valid_qc_sig();
+
+   // if received_qc_sig does not have value, consider valid_qc_sig_from_open only
+   if( !received_qc_sig ) {
+      return std::optional{std::move(valid_qc_sig_from_open)};
+   }
+
+   // Both received_qc_sig and valid_qc_sig_from_open have value. Compare them and select a better one.
+   // Strong beats weak. Tie-break by received_qc_sig, strong beats weak
+   const bool use_qc_sig = received_qc_sig->is_strong() == valid_qc_sig_from_open.is_strong() ? received_qc_sig->is_strong() : false;
+   if (use_qc_sig) {
+      return std::optional{qc_sig_t{ *received_qc_sig }};
+   }
+   return std::optional{qc_sig_t{ std::move(valid_qc_sig_from_open) }};
+}
+
+void open_qc_sig_t::set_received_qc_sig(const qc_sig_t& qc) {
+   std::lock_guard g(*_mtx);
+   received_qc_sig = qc;
+}
+
+bool open_qc_sig_t::received_qc_sig_is_strong() const {
+   std::lock_guard g(*_mtx);
+   return received_qc_sig && received_qc_sig->is_strong();
+}
+
+bool open_qc_sig_t::is_quorum_met_no_lock() const {
+   return is_quorum_met(pending_state);
+}
+
+std::optional<qc_t> open_qc_t::get_best_qc(block_num_type block_num) const {
+   std::optional<qc_sig_t> active_best_qc = active_policy_sig.get_best_qc();
+
+   if (!pending_policy_sig) { // consider only active
+      if (active_best_qc) {
+         return std::optional{qc_t{ block_num, std::move(*active_best_qc), {} }};
+      }
+      return {};
+   }
+   // consider both
+   std::optional<qc_sig_t> pending_best_qc = pending_policy_sig->get_best_qc();
+   if (pending_best_qc) {
+      return std::optional{qc_t{ block_num, std::move(*active_best_qc), std::move(*pending_best_qc) }};
+   }
+   return {};
```

Comment:
I think it could be simplified to:

```suggestion
   std::optional<qc_sig_t> active_best_qc = active_policy_sig.get_best_qc();
   if (!active_best_qc)
      return {};

   if (pending_policy_sig) 
      return std::optional{qc_t{ block_num, std::move(*active_best_qc), pending_policy_sig->get_best_qc() });

   return std::optional{qc_t{ block_num, std::move(*active_best_qc), {} }};
```

Reply:
If there is a `pending_policy_sig` then it must return a best_qc or the pair does not have a best_qc. But it can be simpler in the same spirit of what you have.

Reply:
If there is no `active_best_qc`, regardless of whether there is a `pending_policy_sig` or not, a qc will not be achieved, right?

Reply:
Correct.